repository/copier.go

```diff
@@ -4,107 +4,191 @@ import (
+// opening files, opening file writers, renaming, etc.
+type Fs interface {
+	// Open a file reader.
+	Open(string) (io.ReadCloser, error)
+	// WriteTo a file. It returns a file writer.
+	WriteTo(string) (io.WriteCloser, error)
+	// Rename atomically a file from one path to another.
+	Rename(src, dst string) error
+	// DeleteIfExists deletes a file only if it exists.
+	DeleteIfExists(string) error
+	// Base returns the base path of the filesystem to write files to.
+	Base() string
+	// TempDir returns the base path for temporary directories.
+	TempDir() string
 }
 
-// Copier is in charge either to obtain a file from the Remote filesystem implementation,
-// or send it from local.
-type Copier interface {
-	CopyFromRemote(src, dst string, localFs billy.Filesystem) error
-	CopyToRemote(src, dst string, localFs billy.Filesystem) error
+// Copier is in charge of copying files from a local filesystem to the remote
+// one and vice-versa. It can optionally bucket files on the remote filesystem.
+type Copier struct {
+	remote     Fs
+	local      *localFs
+	bucketSize int
 }
 
-// NewLocalCopier returns a Copier using as a remote a Billy filesystem
-func NewLocalCopier(fs billy.Filesystem, bucket int) Copier {
-	return &LocalCopier{fs, bucket}
+// NewCopier creates a new copier.
+func NewCopier(local billy.Filesystem, remote Fs, bucketSize int) *Copier {
+	return &Copier{remote: remote, local: &localFs{local}, bucketSize: bucketSize}
 }
 
-type LocalCopier struct {
-	fs         billy.Filesystem
-	bucketSize int
+// Local returns the local filesystem of the copier.
+func (c *Copier) Local() billy.Filesystem {
+	return c.local.Filesystem
 }
 
-func (c *LocalCopier) CopyFromRemote(src, dst string, localFs billy.Filesystem) error {
-	bSrc := addBucketName(src, c.bucketSize)
-	return c.copyFile(c.fs, localFs, bSrc, dst)
-}
+// CopyToRemote copies a file from the local filesystem to the remote one.
+func (c *Copier) CopyToRemote(ctx context.Context, src, dst string) error {
+	dstCopy := filepath.Join(c.remote.TempDir(), dst+".copy")
+	dst = filepath.Join(
+		c.remote.Base(),
+		addBucketName(dst, c.bucketSize),
+	)
 
-func (c *LocalCopier) CopyToRemote(src, dst string, localFs billy.Filesystem) error {
-	bDst := addBucketName(dst, c.bucketSize)
-	return c.copyFile(localFs, c.fs, src, bDst)
-}
+	lf, err := c.local.Open(filepath.Join(c.local.Base(), src))
+	if err != nil {
+		if os.IsNotExist(err) {
+			return nil
+		}
+		return err
+	}
 
-func (c *LocalCopier) copyFile(fromFs, toFs billy.Filesystem, from, to string) (err error) {
-	src, err := fromFs.Open(from)
-	if os.IsNotExist(err) {
-		return nil
+	defer checkClose(lf, &err)
+
+	w, err := c.remote.WriteTo(dstCopy)
+	if err != nil {
+		return err
 	}
 
+	// Delete temporary file in case the process is stopped while copying
+	defer func() {
+		_ = w.Close()
+		_ = c.remote.DeleteIfExists(dstCopy)
+	}()
+
+	err = copy(ctx, w, lf)
+
+	checkClose(w, &err)
```

Comment:
`checkClose` does it close when it's not closed? Because we already close the `w` in `defer func`.

Reply:
That defer ignores all errors because it's only meant as a safety net in case the execution stopped for any reason and there's a panic. I could remove the defer, though. `w` is closed here and the file deleted later on, anyway.